src/types/thread_based_random_provider.rs

```diff
@@ -0,0 +1,63 @@
+use rand::Rng;
+
+pub trait RandomProvider {
+    fn generate(&self, base: &ThreadBasedRandomProvider) -> u32;
+}
+
+pub struct ThreadBasedRandomProvider {
+    pub base_value: u32,
+    pub dices: Vec<[u32; 2]>,
+}
+
+impl RandomProvider for ThreadBasedRandomProvider {
+    fn generate(&self, _base: &ThreadBasedRandomProvider) -> u32 {
+        let mut rng = rand::thread_rng();
+        let mut result = self.base_value;
+        for &dice in &self.dices {
```

Comment:
Random provider should expose method `generate`, which takes in **`GenerationBase`**, base_value and dices ownership should be completely logically separated from RandomProvider - RandomProvider does not own GenerationBase, it has a method taking in GenerationBase
```suggestion
pub struct ThreadBasedRandomProvider {
}

impl RandomProvider for ThreadBasedRandomProvider {
    fn generate(&self, base: &GenerationBase) -> u32 {
        let mut rng = rand::thread_rng();
        let mut result = base.base_value;
        for &dice in base.dices {
```